File: Java Projects/Numeric Matrix Processor (Medium)/Stage 3/Processor.java
package processor;

import java.util.Scanner;
import java.util.Arrays;

public class Processor {
    private Scanner scanner;

    public Processor() {
        scanner = new Scanner(System.in);
        Matrix.setScanner(scanner);
    }

    public void execute() {
        executionLoop:
        while (true) {
            System.out.println("1. Add matrices");
            System.out.println("2. Multiply matrix by a constant");
            System.out.println("3. Multiply matrices");
            System.out.println("0. Exit");
            System.out.print("Your choice: ");
            String command = scanner.nextLine();
            try {
                switch (command) {
                    case "1":
                        addMatrices();
                        break;
                    case "2":
                        scalarMultiplication();
                        break;
                    case "3":
                        matrixMultiplication();
                        break;
                    case "0":
                        break executionLoop;
                }
            } catch (IllegalArgumentException e) {
                System.out.println("The operation cannot be performed.");
            }
            System.out.println();
        }
    }

    private void addMatrices() throws IllegalArgumentException {
        System.out.print("Enter size of first matrix: ");
        int[] dims = getDimensions();
        System.out.println("Enter first matrix: ");
        Matrix a = Matrix.getMatrix(dims[0], dims[1]);
        System.out.print("Enter size of second matrix: ");
        dims = getDimensions();
        System.out.println("Enter second matrix: ");
        Matrix b = Matrix.getMatrix(dims[0], dims[1]);
        System.out.println("The result is: ");
        Matrix.add(a, b).printMatrix();
    }

    private void scalarMultiplication() {
        System.out.print("Enter size of matrix: ");
        int[] dims = getDimensions();
        System.out.println("Enter matrix: ");
        Matrix a = Matrix.getMatrix(dims[0], dims[1]);
        System.out.print("Enter constant: ");
        double constant = Double.parseDouble(scanner.next());
        System.out.println("The result is: ");
        Matrix.scalarMultiply(constant, a).printMatrix();
    }

    private void matrixMultiplication() throws IllegalArgumentException {
        System.out.print("Enter size of first matrix: ");
        int[] dims = getDimensions();
        System.out.println("Enter first matrix: ");
        Matrix a = Matrix.getMatrix(dims[0], dims[1]);
        System.out.print("Enter size of second matrix: ");
        dims = getDimensions();
        System.out.println("Enter second matrix: ");
        Matrix b = Matrix.getMatrix(dims[0], dims[1]);
        System.out.println("The result is: ");
        Matrix.multiply(a, b).printMatrix();
    }

    private int[] getDimensions() {
        return Arrays.stream(scanner.nextLine().split(" ")).mapToInt(Integer::parseInt).toArray();
    }
}

File: Python Projects/Numeric Matrix Processor (Medium)/Stage 3/processor.py
from Matrix import Matrix


def load_matrix(n_rows):
    fields = [[int(float(i)) if float(i).is_integer() else float(i) for i in input().split()] for _ in range(n_rows)]
    return Matrix(fields)


def add_matrices():
    a, b = load_two_matrices()
    return a + b


def scalar_multiply():
    row = int(input("Enter size of matrix:").split()[0])
    print("Enter matrix:")
    matrix = load_matrix(row)
    scalar = float(input("Enter constant: "))
    scalar = int(scalar) if scalar.is_integer() else scalar
    return matrix * scalar


def matrix_multiply():
    a, b = load_two_matrices()
    return a @ b


def load_two_matrices():
    rows = int(input("Enter size of first matrix:").split()[0])
    print("Enter first matrix:")
    a = load_matrix(rows)
    rows = int(input("Enter size of second matrix:").split()[0])
    print("Enter second matrix:")
    b = load_matrix(rows)
    return a, b


def menu():
    while True:
        print("1. Add matrices")
        print("2. Multiply matrix by a constant")
        print("3. Multiply matrices")
        print("0. Exit")
        choice = int(input("Your choice: "))
        if choice == 0:
            break
        elif choice == 1:
            try:
                result = add_matrices()
                print("The result is: ")
                print(result)
            except Exception:
                print("The operation cannot be performed")
        elif choice == 2:
            result = scalar_multiply()
            print("The result is: ")
            print(result)
        elif choice == 3:
            try:
                result = matrix_multiply()
                print("The result is: ")
                print(result)
            except Exception:
                print("The operation cannot be performed")
        print()


menu()